### PR TITLE
Take advantage of non-panicking insns for arithmetic on int subtypes

### LINKED ISSUE #899: Take advantage of non-panicking insns for arithmetic on int subtypes
Use limited range of int subtypes to determine when there is no possibility of overflow of arithmetic operations.

In the case of 32-bit integer operations other than division (we can have divide by 0 here) we can use `IntNoPanicArithmeticBinaryInsn` instead or `IntArithmeticBinaryInsn`.

 This is a front-end optimization because the panicking instructions are different.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Take advantage of non-panicking insns for arithmetic on int subtypes
Resolves #899

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front/expr.bal [hunk 1]
```diff
@@ -574,6 +580,29 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
...
+function isNonPanickingIntOp(bir:Operand lhs, bir:Operand rhs, bir:ArithmeticBinaryOp op) returns boolean {
+    t:IntSubtypeConstraints? lhsConstraints = t:intSubtypeConstraints(lhs.semType);
+    t:IntSubtypeConstraints? rhsConstraints = t:intSubtypeConstraints(rhs.semType);
+    if lhsConstraints == () || rhsConstraints == () {
+        return false;
+    }
+    match op {
+        "+"|"-" => {
+            // largest type that can't overflow is unsigned32 (positive) and signed32 (negative)
+            return lhsConstraints.max <= 0xffffffff && rhsConstraints.max <= 0xffffffff &&
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Use predefined constants in lang.int e.g. `int:UNSIGNED32_MAX_VALUE`.

In the match expression, set two variables noPanicMin and noPanicMax. Then check them against each constraint in a loop at the end.

```
foreach var c in constraints {
    if c.min < noPanicMin || c.max > noPanicMax {
        return false
    }
}
return true;
```

File: compiler/modules/front/expr.bal [hunk 2]
```diff
@@ -574,6 +580,29 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
...
+                   lhsConstraints.min >= -0x80000000 && rhsConstraints.min >= -0x80000000;
+        }
+        "*" => {
+            // largest type that can't overflow is unsigned16 (positive) and signed16 (negative)
+            return lhsConstraints.max <= 0xffff && rhsConstraints.max <= 0xffff &&
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Really? Can two int:Signed32 values overflow? Which?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/expr.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: Sorry I didn't notice that there is a typo in the [spec](https://ballerina.io/spec/lang/2021R1/#built-in_subtypes) (positive side have one digit more than the negative side).

Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: created issue #1038

File: compiler/modules/front/expr.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: fixed

File: compiler/modules/front/expr.bal [hunk 3]
```diff
@@ -510,7 +510,13 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
             return { result: { value, semType: resultType }, block: bb };
         }
         result = cx.createTmpRegister(t:INT, pos);
-        bir:IntArithmeticBinaryInsn insn = { op, pos, operands, result };
+        bir:Insn insn;
+        if isNonPanickingIntOp(operands[0], operands[1], op) {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: I wrote a comment but it disappeared. Write like this

```
var name = intArithmeticOpNeverPanics(op, operands);
insn = { op, pos, name, operand, result };
```

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/expr.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: Fixed. Not sure if this is a bug but if I use name directly as `insn = { op, pos, name, operand, result };` (without if) compiler complains saying `a type compatible with mapping constructor expressions not found in type 'wso2/nballerina.bir:0.1.0:Insn'`

Comment (jclark) [reply compiler/modules/front/expr.bal]: Yes, I think that is a JBUG: I seem to remember coming across something similar.

File: compiler/modules/front/expr.bal [hunk 4]
```diff
@@ -574,6 +581,38 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
     return { result, block: bb };
 }
 
+function intArithmeticOpNeverPanics(bir:ArithmeticBinaryOp op, bir:Operand[] operands) returns bir:INSN_INT_NO_PANIC_ARITHMETIC_BINARY|bir:INSN_INT_ARITHMETIC_BINARY {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: A function with this name needs to return boolean.

File: compiler/modules/front/expr.bal [hunk 5]
```diff
@@ -510,7 +510,14 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
             return { result: { value, semType: resultType }, block: bb };
         }
         result = cx.createTmpRegister(t:INT, pos);
-        bir:IntArithmeticBinaryInsn insn = { op, pos, operands, result };
+        var name = intArithmeticOpNeverPanics(op, operands);
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Add JBUG comment

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/expr.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: fixed

File: compiler/modules/front/expr.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: fixed

**Pull request merged**